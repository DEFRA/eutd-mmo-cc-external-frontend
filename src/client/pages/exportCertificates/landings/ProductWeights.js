import React from 'react';
import { GridRow, GridCol, Table, Header } from 'govuk-react';
import { WeightInput } from '../../../components/elements/WeightInput';
import { useTranslation } from 'react-i18next';

function ProductWeights(props) {
  const { species, onWeightChange, errors } = props;
  const {t} = useTranslation();
  const totalExportWeight = species.reduce(
    (acc, curr) => acc + (Number(curr.exportWeight) || 0),
    0
  );

const splitErrorParams = (error) => {
    const isErrorWithParams = error.includes('-');
    const splitError = error.split('-');
    return isErrorWithParams ? t(splitError[0], {dynamicValue :splitError[1],param1 :splitError[2],param2 :splitError[3],param3 :splitError[4]}) : t(error);
  };

 
  return (
    <div>
      <Header level="2">{t('ccAddLandingProductWeightHeader')}</Header>
      <GridRow>
        <GridCol>
          <Table
            id="species-table"
            head={
              <Table.Row>
                <Table.CellHeader>{t('ccAddLandingProductLabel')}</Table.CellHeader>
                <Table.CellHeader>{t('ccAddLandingExportWeightColLabel')}</Table.CellHeader>
              </Table.Row>
            }
            body={
              <>
                {species.map((landing, index) => {
                  const id = `weights.${index}.exportWeight`;
                  
                  const error = errors.find((err) => err.targetName === id);
                
                  return (
                    <Table.Row id={`species_${landing.speciesId}`} key={index}>
                      <Table.Cell>{landing.speciesLabel}</Table.Cell>
                      <Table.Cell>
                        <WeightInput
                          id={id}
                          error={error ?  splitErrorParams(error.text) : ''}
                          value={landing.exportWeight}
                          onChange={(event) => {
                            event.preventDefault();

                            onWeightChange(
                              landing.speciesId,
                              event.target.value
                            );
                          }}
                        />
                      </Table.Cell>
                    </Table.Row>
                  );
                })}
                <Table.Row>
                  <Table.Cell>
                    <strong>{t('ccAddLandingTotalExportWeight')}</strong>
                  </Table.Cell>
                  <Table.Cell id="totalExportWeight">
                    <strong>
                      {Number.isInteger(totalExportWeight)
                        ? totalExportWeight
                        : totalExportWeight.toFixed(2)}
                      {t('ccDirectLandingProductWeightTableExportWeightInputUnit')}
                    </strong>
                  </Table.Cell>
                </Table.Row>
              </>
            }
          />
        </GridCol>
      </GridRow>
    </div>
  );
}

export default ProductWeights;
