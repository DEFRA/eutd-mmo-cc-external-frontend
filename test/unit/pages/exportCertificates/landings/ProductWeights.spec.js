import ProductWeights from '../../../../../src/client/pages/exportCertificates/landings/ProductWeights';
import { mount } from 'enzyme';
import React from 'react';
import { render } from '@testing-library/react';

describe('ProductWeights', () => {
  let wrapper;

  const species = [
    {
      speciesId: '1',
      speciesLabel: 'Cod',
      exportWeight: 100,
    },
    {
      speciesId: '2',
      speciesLabel: 'Herring',
      exportWeight: 200,
    },
  ];

  const errors = [
    {
      targetName: 'weights.1.exportWeight',
      text: 'error!'
    }
  ];

  const onChange = jest.fn();

  beforeAll(() => {
    wrapper = mount(
      <ProductWeights species={species} onWeightChange={onChange} errors={errors} />
    );
  });

  it('will render', () => {
    expect(wrapper).not.toBeNull();
    expect(wrapper.text()).toContain('Product weights');
  });

  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
  
  it('will show the correct total export weight', () => {
    expect(wrapper.exists('#totalExportWeight')).toBe(true);
    expect(wrapper.find('#totalExportWeight').at(0).text()).toBe('300kg');
  });

  it('will call onChange when a weight is updated', () => {
    expect(onChange).not.toHaveBeenCalled();

    const first = wrapper.find('input[name="weight"]').first();

    first.simulate('change', { target: { name: 'weight', value: '1000' } });

    expect(onChange).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledWith('1', '1000');
  });

  it('will count invalid weights as 0', () => {
    const speciesWithInvalidWeight = [
      {
        speciesId: '1',
        speciesLabel: 'Cod',
        exportWeight: 100.10,
      },
      {
        speciesId: '2',
        speciesLabel: 'Herring',
        exportWeight: '23,3333',
      }
    ];

    const wrapperWithInvalidWeight = mount(
      <ProductWeights
        species={speciesWithInvalidWeight}
        onWeightChange={onChange}
        errors={errors}
      />
    );

    expect(wrapperWithInvalidWeight.exists('#totalExportWeight')).toBe(true);
    expect(
      wrapperWithInvalidWeight.find('#totalExportWeight').at(0).text()
    ).toBe('100.10kg');
  });

  it('will set a unique id for every weight input', () => {
    expect(wrapper.find('WeightInput').at(0).prop('id')).toBe('weights.0.exportWeight');
    expect(wrapper.find('WeightInput').at(1).prop('id')).toBe('weights.1.exportWeight');
  });

  it('will show an error', () => {
    expect(wrapper.find('WeightInput').at(0).prop('error')).toBe('');
    expect(wrapper.find('WeightInput').at(1).prop('error')).toBe('error!');
  });
});
