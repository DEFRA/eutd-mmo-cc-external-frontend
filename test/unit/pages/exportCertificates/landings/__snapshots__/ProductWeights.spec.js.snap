// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductWeights should take a snapshot of the whole page 1`] = `
<div>
  <div>
    <h2
      class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
    >
      Product weights
    </h2>
    <div
      class="css-1lzr9ic-StyledRow e4u1u2g0"
    >
      <div
        class="css-1w80l5h-StyledColumn e4u1u2g0"
      >
        <table
          class="css-1uqv7s3-StyledTable e4u1u2g0"
          id="species-table"
        >
          <caption
            class="css-gzlitc-Caption e172z3n50"
          />
          <thead
            class="css-1iuscil-TableHeader e1tbxulj0"
          >
            <tr
              class="css-7u9vzt-RowInner e1bh3ka20"
            >
              <th
                class="css-12e3svi-CellHeader e1sike1z0"
              >
                Product
              </th>
              <th
                class="css-12e3svi-CellHeader e1sike1z0"
              >
                Export weight
              </th>
            </tr>
          </thead>
          <tbody
            class="css-x1p7io-TableBody eo9ngsf0"
          >
            <tr
              class="css-7u9vzt-RowInner e1bh3ka20"
              id="species_1"
            >
              <td
                class="css-14ci18q-CellInner edqrv1d0"
              >
                Cod
              </td>
              <td
                class="css-14ci18q-CellInner edqrv1d0"
              >
                <div
                  class="govuk-form-group"
                  name="exportWeight"
                >
                  <label
                    class="govuk-label-weight-input e1vcssft0 css-137pbp-StyledLabel-StyledHoc e4u1u2g0"
                    for="weight"
                  />
                  <div
                    class="govuk-input__wrapper"
                  >
                    <input
                      class="form-control weight-input"
                      id="weights.0.exportWeight"
                      name="weight"
                      spellcheck="false"
                      style="margin-right: 0px;"
                      type="text"
                      value="100"
                    />
                    <div
                      aria-hidden="true"
                      class="govuk-input__suffix"
                    >
                      kg
                    </div>
                  </div>
                </div>
              </td>
            </tr>
            <tr
              class="css-7u9vzt-RowInner e1bh3ka20"
              id="species_2"
            >
              <td
                class="css-14ci18q-CellInner edqrv1d0"
              >
                Herring
              </td>
              <td
                class="css-14ci18q-CellInner edqrv1d0"
              >
                <div
                  class="govuk-form-group govuk-form-group-error"
                  name="exportWeight"
                >
                  <label
                    class="govuk-label-weight-input e1vcssft0 css-137pbp-StyledLabel-StyledHoc e4u1u2g0"
                    for="weight"
                  />
                  <div
                    class="error-message"
                  >
                    error!
                  </div>
                  <div
                    class="govuk-input__wrapper"
                  >
                    <input
                      class="form-control form-control-error weight-input"
                      id="weights.1.exportWeight"
                      name="weight"
                      spellcheck="false"
                      style="margin-right: 0px;"
                      type="text"
                      value="200"
                    />
                    <div
                      aria-hidden="true"
                      class="govuk-input__suffix"
                    >
                      kg
                    </div>
                  </div>
                </div>
              </td>
            </tr>
            <tr
              class="css-7u9vzt-RowInner e1bh3ka20"
            >
              <td
                class="css-14ci18q-CellInner edqrv1d0"
              >
                <strong>
                  Total export weight
                </strong>
              </td>
              <td
                class="css-14ci18q-CellInner edqrv1d0"
                id="totalExportWeight"
              >
                <strong>
                  300
                  kg
                </strong>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
